server/src/main/java/org/opensearch/common/blobstore/AsyncMultiStreamEncryptedBlobContainer.java

```diff
@@ -108,4 +106,69 @@ public InputStreamContainer provideStream(int partNumber) throws IOException {
         }
 
     }
+
+    static class DecryptedReadContext<T, U> extends ReadContext {
+
+        private final U cryptoContext;
+        private final CryptoHandler<T, U> cryptoHandler;
+        private final long fileSize;
+
+        public DecryptedReadContext(ReadContext readContext, CryptoHandler<T, U> cryptoHandler) {
+            super(readContext);
+            this.cryptoHandler = cryptoHandler;
+            try {
+                this.cryptoContext = this.cryptoHandler.loadEncryptionMetadata(null);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            this.fileSize = this.cryptoHandler.estimateDecryptedLength(cryptoContext, readContext.getBlobSize());
+        }
+
+        @Override
+        public long getBlobSize() {
+            return fileSize;
+        }
+
+        @Override
+        public List<InputStreamContainer> getPartStreams() {
+            return super.getPartStreams().stream().map(this::decrpytInputStreamContainer).collect(Collectors.toList());
+        }
+
+        private InputStreamContainer decrpytInputStreamContainer(InputStreamContainer inputStreamContainer) {
+            long startOfStream = inputStreamContainer.getOffset();
+            long endOfStream = startOfStream + inputStreamContainer.getContentLength();
+            DecryptedRangedStreamProvider rangedStreamProvider = cryptoHandler.createDecryptingStreamOfRange(
+                cryptoContext,
+                startOfStream,
+                endOfStream
+            );
+
+            InputStream decryptedStream = cryptoHandler.createDecryptingStream(inputStreamContainer.getInputStream());
```

Comment:
You already created a ranged decrypting stream. Decryption should happen only once and hence, there should only be one wrapping decrypting stream. You can remove ranged stream and just use this.

Reply:
Updated. Thanks